File: LETS-GET-A-JOB/scripts/fix-n8n-workflows.js
#!/usr/bin/env node

/**
 * Fix n8n Workflow Input Configurations
 *
 * This script fixes the AI Agent input in all 4 workflows to properly
 * read data from the webhook node.
 *
 * Problem: AI Agent is trying to read $json.jobUrl but webhook data
 * is in $('Webhook').item.json.body.jobUrl
 *
 * Solution: Update the AI Agent node's input expression for each workflow
 */

const { execSync } = require('child_process');
const path = require('path');
const os = require('os');

const DB_PATH = path.join(os.homedir(), '.n8n', 'database.sqlite');

console.log('🔧 n8n Workflow Input Fixer\n');
console.log(`📂 Database: ${DB_PATH}\n`);

const db = new sqlite3.Database(DB_PATH, (err) => {
  if (err) {
    console.error('❌ Error opening database:', err.message);
    process.exit(1);
  }
  console.log('✅ Connected to n8n database\n');
});

// Workflow fixes configuration
const WORKFLOW_FIXES = [
  {
    name: 'Job Desc',
    description: 'Job Description Processing',
    aiAgentInput: '={{ $("Webhook").item.json.body.jobUrl }}',
    explanation: 'Reads jobUrl from webhook POST body'
  },
  {
    name: 'Resume',
    description: 'Resume PDF Processing',
    aiAgentInput: '={{ $("Webhook").item.binary.data }}',
    explanation: 'Reads PDF file from webhook binary upload (field name: data)'
  },
  {
    name: 'Cover Letter',
    description: 'Cover Letter PDF Processing',
    aiAgentInput: '={{ $("Webhook").item.binary.data }}',
    explanation: 'Reads PDF file from webhook binary upload (field name: data)'
  },
  {
    name: 'Eval',
    description: 'ATS Evaluation',
    aiAgentInput: '={{ $("Webhook").item.json.body }}',
    explanation: 'Reads entire body with resume_text, cover_letter_text, job_description'
  }
];

async function getWorkflow(name) {
  return new Promise((resolve, reject) => {
    db.get(
      'SELECT id, name, nodes FROM workflow_entity WHERE name = ?',
      [name],
      (err, row) => {
        if (err) reject(err);
        else resolve(row);
      }
    );
  });
}

async function updateWorkflow(id, nodes) {
  return new Promise((resolve, reject) => {
    db.run(
      'UPDATE workflow_entity SET nodes = ? WHERE id = ?',
      [JSON.stringify(nodes), id],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

async function fixWorkflow(config) {
  console.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`🔍 Fixing: ${config.name} (${config.description})`);

  try {
    // Get workflow
    const workflow = await getWorkflow(config.name);

    if (!workflow) {
      console.log(`❌ Workflow "${config.name}" not found\n`);
      return false;
    }

    // Parse nodes
    const nodes = JSON.parse(workflow.nodes);

    // Find AI Agent node
    const aiAgentNode = nodes.find(n => n.type === '@n8n/n8n-nodes-langchain.agent');

    if (!aiAgentNode) {
      console.log(`❌ AI Agent node not found in workflow\n`);
      return false;
    }

    // Show current input
    const currentInput = aiAgentNode.parameters.text || 'Not set';
    console.log(`   Current input: ${currentInput}`);

    // Update input
    aiAgentNode.parameters.text = config.aiAgentInput;
    console.log(`   New input:     ${config.aiAgentInput}`);
    console.log(`   Explanation:   ${config.explanation}`);

    // Save workflow
    await updateWorkflow(workflow.id, nodes);

    console.log(`✅ Fixed successfully!\n`);
    return true;

  } catch (error) {
    console.log(`❌ Error: ${error.message}\n`);
    return false;
  }
}

async function main() {
  let successCount = 0;
  let failCount = 0;

  for (const config of WORKFLOW_FIXES) {
    const success = await fixWorkflow(config);
    if (success) successCount++;
    else failCount++;
  }

  console.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`\n📊 Summary:`);
  console.log(`   ✅ Fixed: ${successCount}`);
  console.log(`   ❌ Failed: ${failCount}`);
  console.log(`   📝 Total: ${WORKFLOW_FIXES.length}\n`);

  if (successCount === WORKFLOW_FIXES.length) {
    console.log(`🎉 All workflows fixed successfully!`);
    console.log(`\n🔄 Next steps:`);
    console.log(`   1. Restart n8n (if running)`);
    console.log(`   2. Test webhooks using: open scripts/test-n8n-webhooks.html`);
    console.log(`   3. Verify all 4 workflows return 200 OK\n`);
  } else {
    console.log(`⚠️  Some workflows failed to fix. Check errors above.\n`);
  }

  db.close();
}

main().catch(err => {
  console.error('❌ Fatal error:', err);
  db.close();
  process.exit(1);
});

File: LETS-GET-A-JOB/scripts/fix-n8n-workflows.py
#!/usr/bin/env python3

"""
Fix n8n Workflow Input Configurations

This script fixes the AI Agent input in all 4 workflows to properly
read data from the webhook node.

Problem: AI Agent is trying to read $json.jobUrl but webhook data
is in $('Webhook').item.json.body.jobUrl

Solution: Update the AI Agent node's input expression for each workflow
"""

import sqlite3
import json
import os
from pathlib import Path

DB_PATH = Path.home() / '.n8n' / 'database.sqlite'

# Workflow fixes configuration
WORKFLOW_FIXES = [
    {
        'name': 'Job Desc',
        'description': 'Job Description Processing',
        'ai_agent_input': '={{ $("Webhook").item.json.body.jobUrl }}',
        'explanation': 'Reads jobUrl from webhook POST body'
    },
    {
        'name': 'Resume',
        'description': 'Resume PDF Processing',
        'ai_agent_input': '={{ $("Webhook").item.binary.data }}',
        'explanation': 'Reads PDF file from webhook binary upload (field name: data)'
    },
    {
        'name': 'Cover Letter',
        'description': 'Cover Letter PDF Processing',
        'ai_agent_input': '={{ $("Webhook").item.binary.data }}',
        'explanation': 'Reads PDF file from webhook binary upload (field name: data)'
    },
    {
        'name': 'Eval',
        'description': 'ATS Evaluation',
        'ai_agent_input': '={{ $("Webhook").item.json.body }}',
        'explanation': 'Reads entire body with resume_text, cover_letter_text, job_description'
    }
]

def fix_workflow(conn, config):
    """Fix a single workflow's AI Agent input"""
    print(f"━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"🔍 Fixing: {config['name']} ({config['description']})")
    
    try:
        # Get workflow
        cursor = conn.cursor()
        cursor.execute(
            'SELECT id, name, nodes FROM workflow_entity WHERE name = ?',
            (config['name'],)
        )
        row = cursor.fetchone()
        
        if not row:
            print(f"❌ Workflow \"{config['name']}\" not found\n")
            return False
        
        workflow_id, name, nodes_json = row
        
        # Parse nodes
        nodes = json.loads(nodes_json)
        
        # Find AI Agent node
        ai_agent_node = None
        for node in nodes:
            if node.get('type') == '@n8n/n8n-nodes-langchain.agent':
                ai_agent_node = node
                break
        
        if not ai_agent_node:
            print(f"❌ AI Agent node not found in workflow\n")
            return False
        
        # Show current input
        current_input = ai_agent_node.get('parameters', {}).get('text', 'Not set')
        print(f"   Current input: {current_input}")
        
        # Update input
        if 'parameters' not in ai_agent_node:
            ai_agent_node['parameters'] = {}
        ai_agent_node['parameters']['text'] = config['ai_agent_input']
        
        print(f"   New input:     {config['ai_agent_input']}")
        print(f"   Explanation:   {config['explanation']}")
        
        # Save workflow
        cursor.execute(
            'UPDATE workflow_entity SET nodes = ? WHERE id = ?',
            (json.dumps(nodes), workflow_id)
        )
        conn.commit()
        
        print(f"✅ Fixed successfully!\n")
        return True
        
    except Exception as e:
        print(f"❌ Error: {str(e)}\n")
        return False

def main():
    print('🔧 n8n Workflow Input Fixer\n')
    print(f'📂 Database: {DB_PATH}\n')
    
    if not DB_PATH.exists():
        print(f'❌ Database not found at {DB_PATH}')
        print('   Make sure n8n is installed and has been run at least once.')
        return
    
    # Connect to database
    conn = sqlite3.connect(str(DB_PATH))
    print('✅ Connected to n8n database\n')
    
    success_count = 0
    fail_count = 0
    
    for config in WORKFLOW_FIXES:
        if fix_workflow(conn, config):
            success_count += 1
        else:
            fail_count += 1
    
    conn.close()
    
    print(f"━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"\n📊 Summary:")
    print(f"   ✅ Fixed: {success_count}")
    print(f"   ❌ Failed: {fail_count}")
    print(f"   📝 Total: {len(WORKFLOW_FIXES)}\n")
    
    if success_count == len(WORKFLOW_FIXES):
        print(f"🎉 All workflows fixed successfully!")
        print(f"\n🔄 Next steps:")
        print(f"   1. Restart n8n (if running)")
        print(f"   2. Test webhooks using: open scripts/test-n8n-webhooks.html")
        print(f"   3. Verify all 4 workflows return 200 OK\n")
    else:
        print(f"⚠️  Some workflows failed to fix. Check errors above.\n")

if __name__ == '__main__':
    main()

